analog_forecast.py

```diff
@@ -0,0 +1,297 @@
+    
+    return rmse_da
+
+
+def run_rmse_over_time(da, ref_date, window):
+    """Get the RMSE between da at ref_time and every other preceding time slice. Calls compute() to process with dask.
+    
+    Args:
+        da (xarray.DataArray): ERA5 dataArray, with time, latitude, longitude axes
+        ref_date (str): reference date to compare all other preceding time slices against
+        window (str): method for temporal search, passed to get_search_da
+        
+    Returns:
+        rmse_da (xarra.DataArray): data array with an RMSE computed for each date preceding ref_date
+    """
+    ref_da = da.sel(time=ref_date).squeeze()
+    search_da = get_search_da(da, ref_date, window)
+    rmse_da = rmse(search_da, ref_da)
+    
+    return rmse_da
+
+
+def take_analogs(error_da, buffer, ref_date, n=5):
+    """Take the top n times from error_da ('top' == smallest error) with constraints on how close in time these top times may be.
+    
+    Args:
+        error_da (xarray.DataArray): error values indexed by time
+        buffer (int): number of days to buffer each top time for excluding subsequent times
+        ref_date (str): 
+        n (int): number of analogs to keep
+        
+    Returns:
+        analogs (xarray.DataArray): top n analogs taken from error_da
+    """
+    # sort ascending
+    error_da = error_da.sortby(error_da)
+    # iterate over list of top analogs and create a window that others must not overlap, lest they be discarded
+    analog_buffer_ranges = []
+    analog_times = []
+    td_buffer = pd.to_timedelta(buffer, "d")
+    for t in error_da.time.values:
+        if not any([t in dr for dr in analog_buffer_ranges]):
+            analog_buffer_ranges.append(pd.date_range(t - td_buffer, t + td_buffer))
+            analog_times.append(t)
+        if len(analog_buffer_ranges) == 5:
+            break
+
+    analogs = error_da.sel(time=analog_times)
+
+    return analogs
+
+
+def find_analogs(da, ref_date, print_analogs=False):
+    """Find the analogs.
+    
+    Args:
+        da (xarray.DataArray): data array of ERA5 data (likely already subset to area of interest)
+        ref_date (str): reference date in formate YYYY-mm-dd
+        print_analogs (bool): print the top 5 analogs and scores
+        
+    Returns:
+        analogs (xarray.DataArray): data array of RMSE values and dates for 5 best analogs
+    """
+    # compute RMSE between ref_date and all preceding dates 
+    #  for the specified variable and spatial domain
+    rmse_da = run_rmse_over_time(da, ref_date, "any")
+    varname = da.name
+    
+    # subset to first 5 analogs for now
+    # impose restrictions on temporal proximity for analog based on variable.
+    #  if SST, then we use an exclusion buffer of 6 months, atmospheric vars use 30 days
+    if varname in ["t2m", "msl", "z"]:
+        buffer = 30
+    elif varname in ["sst"]:
+        buffer = 180
+    analogs = take_analogs(rmse_da, buffer, ref_date, 5)
+    
+    if print_analogs:
```

Comment:
limit of 5 analogues is also hard coded here...consider using an "n" parameter like the ```take_analogs``` function above?